Conflict Resolution implementation

itegulov opened the pull request:
Sequent and concurrent versions for CR algorithm were implemented.
Also, contains some tests from TPTP library with simple time measurement.

itegulov committed:
Add Sequent structures necessary for propositional logic

itegulov committed:
Enrich UnitSequent and VarSeqSequent

itegulov committed:
Add type aliases for algorithm.prover package

itegulov committed:
Add CNF structure with two-watched literal

itegulov committed:
Add documentation for Cnf

itegulov committed:
Introduce DecisionLevel, ConflictAnalyser and LiteralChooser

Also added simple implementations for them

itegulov committed:
Refactor Cnf::assignLiteral

It's nearly beautiful now

itegulov committed:
Implement CDCL algorithm

Also added some tests for it

itegulov committed:
Introduce immutable Cnf as an input for CDCL

itegulov committed:
Rename Cnf to CNF and assessment to assignment

itegulov committed:
Replace redundant classes with some implicit conversions

They didn't provide any new functionality and just added some
type safety, though it will be pretty hard to use it. So I replaced
it with some neat implicit conversions.

itegulov committed:
Merge branch 'develop' of https://github.com/Paradoxika/Skeptik into cr-implementation

itegulov committed:
Implement Unit-Propagation Resolution rule

itegulov committed:
Move some general implicits to prover package

itegulov committed:
Implement Conflict rule

itegulov committed:
Remove redundant implicit variables

itegulov committed:
Implement resolution procedure for FOL with BFS strategy

itegulov committed:
Ensure that combinations generation is lazy

itegulov committed:
Prettify code: add comments and little refactor

itegulov committed:
New clauses shouldn't be contained in previous levels

itegulov committed:
Introduce an ancestor-based idea

We can make necessary decisions to support invariant: at each level
we derive new clauses using older clauses and in total use every ancestor.

Further research of this approach is necessary.

itegulov committed:
Move resolving to separate method

itegulov committed:
Implement template for deciding in ancestor-based idea

itegulov committed:
Fix some bugs, prettify code

itegulov committed:
Implement draft version of ancestor-based idea

It does something "nearly" reasonable, but there are some implementation
bugs: e.g. if it decides (P(a) |= ) on some step, on the next step it can fail
to ensure trueness of (P(a) |= P(b)).

Code is very dirty. If you're trying to understand it -- better wait for new
commits: I'll certainly clean this up and add some documentation with comms.

itegulov committed:
Fix some bugs in CR implementation

Now it produces all new possible conclusions and add them. Also it detects a
conflict now. But it doesn't learn new clause for now (just halts).

Code is still very dirty. If you're trying to understand it -- better wait for new
commits: I'll certainly clean this up and add some documentation with comms.

itegulov committed:
Remove debug output and consider already satisfied clauses as used

Code is still very dirty. If you're trying to understand it -- better wait for new
commits: I'll certainly clean this up and add some documentation with comms.

itegulov committed:
Add very simple clause learning and backtracking

It even works for simple cases, but testing on more complex examples is
required.

Code is still very dirty. If you're trying to understand it -- better wait for new
commits: I'll certainly clean this up and add some documentation with comms.

itegulov committed:
Clause learning uses instantiations of decisions now

Also randomized interesting clause choosing

Code is still very dirty. If you're trying to understand it -- better wait for new
commits: I'll certainly clean this up and add some documentation with comms.

itegulov committed:
Add renaming of quantified variables before resolving

Code is still very dirty. If you're trying to understand it -- better wait for new
commits: I'll certainly clean this up and add some documentation with comms.

itegulov committed:
Cleanup and add some comments with documentation for CR

itegulov committed:
Unification in updateClauses also requires renaming

Code is still very dirty. If you're trying to understand it -- better wait for new
commits: I'll certainly clean this up and add some documentation with comms.

itegulov committed:
Switch to SetSequent as Clauses' basis

Code is still very dirty. If you're trying to understand it -- better wait for new
commits: I'll certainly clean this up and add some documentation with comms.

itegulov committed:
Improve backtracking behaviour

Now conflict clause won't be considered as a basic ancestor

Finally solves some basic examples from TPTP! (e.g. SYN-012-1)

Code is still very dirty. If you're trying to understand it -- better wait for new
commits: I'll certainly clean this up and add some documentation with comms.

itegulov committed:
Merge branch 'develop' of https://github.com/Paradoxika/Skeptik into cr-implementation

itegulov committed:
Add AGT001+1 example from TPTP

itegulov committed:
Copied AGT001+2 accidentally: change it to AGT001+1

itegulov committed:
Add occur-check to Martelli Montanari algorithm

itegulov committed:
Leave only most general instantiations in clause

Also fixed bug with resulting with "satisfiable" when there is conflict from the very begining

itegulov committed:
Fix fatal bug in general instances search

itegulov committed:
Merge branch 'develop' of github.com:Paradoxika/Skeptik into cr-implementation

# Conflicts:
#	examples/problems/Axioms/AGT001+0.ax
#	examples/problems/Axioms/AGT001+1.ax
#	examples/problems/Axioms/NUM005+0.ax
#	examples/problems/Axioms/NUM005+1.ax
#	examples/problems/FOF/AGT001+1.fofp
#	src/test/scala/at/logic/skeptik/parser/TPTPTest.scala

itegulov committed:
Remove unnecessary new's

itegulov committed:
Move renaming before unifying in separate method

itegulov committed:
Refactor and add comment for main CR algorithm

itegulov committed:
Use more explicit termination condition

itegulov committed:
Update unifiableUnits if there it's empty

itegulov committed:
Fix bug with checking needed decision after all type of decision

itegulov committed:
Add traversal of implication graph for conflict clause searching

itegulov committed:
Fix substitution bug when searching for conflict clause

itegulov committed:
Introduce some examples of CR work

itegulov committed:
Implication graph should be updated even if the conclusion is already satisfied

itegulov committed:
Big refactor of whole prover package

Move Literal to separate class (finally!)
Change terminology
Improve type safety of internal CR structure: now we resolve literal, not clauses etc
Add a lot of comments and documentation

itegulov committed:
Change Literal's toString

itegulov committed:
Fix some bugs after refactoring, add logging

itegulov committed:
Resolve conflict search bug and resolve invocation bug

Also restructured code a little bit

itegulov committed:
Derive all possible conflict clauses

Also add new test

itegulov committed:
Replace decisions with propagation if it is possible

Also refactored code: removed breaks, explicit while loop and added documentation
with comments for some places.

itegulov committed:
Merge branch 'develop' of github.com:Paradoxika/Skeptik into cr-implementation

itegulov committed:
Fix renaming before unifying: use general substitution for all literals

Also added test covering this case

itegulov committed:
Return different mgu's for left and right parts

itegulov committed:
Include only missing part of substitution when renaming

Also added a lot of tests covering difficult cases

itegulov committed:
Should override decision only if it isn't used to propagate itself

Also fixed some other small bugs and updated tests

itegulov committed:
Reuse decision literals at the same depth

Also add new test with the case where this helps

itegulov committed:
Get rid of loops in implication graph

Intermediate value theorem can be proved now!

itegulov committed:
Refactor renaming mechanism

Also fixed a lot of bugs in it: now substitution compose when trying to
find a conflict clause.

Now some tests don't work as well as they did, because before this fix
some illegal literals were propagated.

itegulov committed:
Move general renaming functions to prover package

itegulov committed:
Also move unifyWithRename to prover package

Updated documentation for it, so it's much clearer what it does now

itegulov committed:
Changed inference rules structure a bit

Also moved them to proper place and fixed some unification bugs
in UnitPropagationResolution

itegulov committed:
Implement Decision and CDCL inference rules

Also adjusted other rules implementation and fixed some compilation errors.

itegulov committed:
Use SeqSequent instead of SetSequent

Also added some useful utility functions

itegulov committed:
Change UPR structure, use SeqSequent in inference rules

Now desired is specified as a literal, not index. Also included possibility to
derive desired clause automatically.

itegulov committed:
Fix unification bug in UPR inference rule

itegulov committed:
CR now returns Proof[SequentProofNode] of ⊥

itegulov committed:
Result of UPR rule should not be desired, but it's instantiation

itegulov committed:
Add documentation for buildProof function

itegulov committed:
Move utility combinations function to package

itegulov committed:
Initial implementation of concurrent CR

itegulov committed:
Propagated literals should overwrite decision literals

itegulov committed:
Fix Akka actors usage bug

There exists a problem with using mutable state while creating futures, who have
access to this mutable state. I fixed this problem temporarily by using futures
only synchronously.

I'll optimise it later.

Also added proof generation instead of isSatisifable result.

itegulov committed:
Add benchmarking for CR and ConcurrentCR

itegulov committed:
Add comments for actor implementation

itegulov committed:
Merge branch 'develop' of github.com:Paradoxika/Skeptik into cr-implementation

Zahnentferner commented on src/main/scala/at/logic/skeptik/algorithm/prover/actors/messages/GetVariables.scala:
```diff
@@ -0,0 +1,6 @@
+/**
+  * @author Daniyar Itegulov
+  */
+case class GetVariables()
```
Can't you use a case object instead of a case class without parameters?

Zahnentferner commented on src/main/scala/at/logic/skeptik/algorithm/prover/conflict/SimpleConflictAnalyser.scala:
```diff
@@ -0,0 +1,14 @@
+  */
+object SimpleConflictAnalyser extends ConflictAnalyser{
+  override def learnConflictClause(levels: Seq[DecisionLevel]): Clause =
+    (Clause.empty /: levels.map(!_.literal)) (_ union _.toClause)
```
This line could be greatly optimized...

Zahnentferner commented on src/main/scala/at/logic/skeptik/algorithm/prover/package.scala:
```diff
@@ -0,0 +1,192 @@
+    def isUnit: Boolean = clause.width == 1
+  }
+
+  implicit class UnitSequent(sequent: SeqSequent) {
```
To improve performance, it is a good idea to make implicit classes extend "AnyVal".  Google about it to learn more.

Zahnentferner commented on src/main/scala/at/logic/skeptik/algorithm/prover/structure/immutable/Cnf.scala:
```diff
@@ -0,0 +1,20 @@
+
+  def -(that: CNF): CNF = new CNF(clauses.filterNot(that.clauses.toSet))
+
+  def toMutableCnf: mutable.CNF = new mutable.CNF(clauses.to[ArrayBuffer])
```
toMutableCNF

Zahnentferner commented on src/main/scala/at/logic/skeptik/proof/sequent/conflictresolution/Decision.scala:
```diff
@@ -0,0 +1,20 @@
+/**
+  * @author Daniyar Itegulov
+  */
+case class Decision(clause: SeqSequent) extends SequentProofNode {
```
Why does Decision take a clause as argument? Why not simply a literal instead?

Zahnentferner commented:
Many thanks, Daniyar!

I left a few line comments here in Github.

Zahnentferner commented:
Also, I noticed that there is no command line interface to use the prover. Could you create one, similar to the "ProofCompressionCLI.scala" ?  It is better to create a new file for the provers, instead of adapting "ProofCompressionCLI.scala", because proving is quite different from compressing proofs...

itegulov commented:
@Ceilican Thanks for your review! I've made necessary changes and fixed all issues you mentioned.

Also, I implemented ProveCLI.scala and registered it in Build.scala as "skeptik-prove" command. But it won't be packed in one-jar target because only one entry point can be used in one jar.

I'll clean up my code some more and will create another PR with all changes after that.

Zahnentferner commented:
Thanks!
